colossus/src/main/scala/colossus/service/LoadBalanceClient.scala

```diff
@@ -0,0 +1,106 @@
+package colossus.service
+
+import colossus.metrics.logging.ColossusLogging
+
+import scala.concurrent.duration._
+import scala.language.higherKinds
+
+trait LBC {
+  def addClient(host: String, port: Int): Unit
+  def removeClient(host: String, port: Int, allInstances: Boolean): Unit
+}
+
+case class LoadBalanceClient[P <: Protocol, M[_], +T <: Sender[P, M], E](
+    hostManager: HostManager,
+    clientFactory: ClientFactory[P, M, T, E])(implicit env: E, async: Async[M])
+    extends Sender[P, M]
+    with LBC
+    with ColossusLogging {
+
+  case class ClientWrapper(client: Sender[P, M], host: String, port: Int)
+
+  private var clients = Seq.empty[ClientWrapper]
+
+  private val retryPolicy = hostManager.retryPolicy
+
+  private var permutations = new PermutationGenerator(clients.map { _.client })
+
+  // first register with the host manager, which acts as the interface between the user and this class
+  hostManager.attachLoadBalancer(this)
+
+  def addClient(host: String, port: Int): Unit = {
+    clients = ClientWrapper(clientFactory(host, port), host, port) +: clients
+    regeneratePermutations()
+  }
+
+  // TODO will this fuck up requests in progress?
```

Comment:
Do you think it is worthwhile to find the next client to use when it is needed instead of upfront? We would just need to keep track of the hosts that have already been used.